src/NHSD.BuyingCatalogue.Ordering.Api/Attributes/LimitAttribute.cs

```diff
@@ -0,0 +1,166 @@
+﻿using System;
+using System.ComponentModel;
+using System.ComponentModel.DataAnnotations;
+using System.Globalization;
+
+namespace NHSD.BuyingCatalogue.Ordering.Api.Attributes
+{
+    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Parameter, AllowMultiple = true)]
+    public sealed class LimitAttribute : ValidationAttribute
+    {
+        public object Value { get; private set; }
+
+        public Type OperandType { get; private set; }
+
+        public LimitType RangeType { get; private set; }
+
+        private Func<object, object> Conversion { get; set; }
+
+        public LimitAttribute(int value, LimitType range)
+            : this()
+        {
+            this.Value = value;
+            this.RangeType = range;
+            this.OperandType = typeof(int);
+        }
+
+        public LimitAttribute(double value, LimitType range)
+            : this()
+        {
+            this.Value = value;
+            this.RangeType = range;
+            this.OperandType = typeof(double);
+        }
+
+        public LimitAttribute(Type type, string value, LimitType range)
+            : this()
+        {
+            this.OperandType = type;
+            this.Value = value;
+            this.RangeType = range;
+        }
+
+        private LimitAttribute()
+            : base(() => "Value was outside the defined limit")
+        {
+        }
+
+        private void Initialize(IComparable value, Func<object, object> conversion)
+        {
+            this.Value = value;
+            this.Conversion = conversion;
+        }
+
+        public override bool IsValid(object value)
+        {
+            // Validate our properties and create the conversion function
+            this.SetupConversion();
+
+            // Automatically pass if value is null or empty. RequiredAttribute should be used to assert a value is not empty.
+            if (value == null)
+            {
+                return true;
+            }
+
+            if (value is string s && String.IsNullOrEmpty(s))
+            {
+                return true;
+            }
+
+            object convertedValue = null;
+
+            try
+            {
+                convertedValue = this.Conversion(value);
+            }
+            catch (FormatException)
+            {
+                return false;
+            }
+            catch (InvalidCastException)
+            {
+                return false;
+            }
+            catch (NotSupportedException)
+            {
+                return false;
+            }
+
+            IComparable val = (IComparable)this.Value;
```

Comment:
edge case I know... but what if this.Value isn't IComparable?

Reply:
It's not possible, the SetupConversion function throws an exception by this point, if it isn't IComparable